labelbox/data/annotation_types/data/tiled_image.py

```diff
@@ -0,0 +1,460 @@
+        """This function slices off the excess pixels that are outside of the bounds.
+        This occurs because only full tiles can be downloaded at a time.
+        """
+
+        def invert_point(pt):
+            # Must have at least 1 pixel for stability.
+            pt = max(pt, 1)
+            # All pixel points are relative to a single tile
+            # So subtracting the tile size inverts the axis
+            pt = pt - self.tile_size
+            return pt if pt != 0 else None
+
+        x_px_end, y_px_end = invert_point(x_px_end), invert_point(y_px_end)
+        return image[y_px_start:y_px_end, x_px_start:x_px_end, :]
+
+    def _validate_num_tiles(self, xstart: float, ystart: float, xend: float,
+                            yend: float, max_tiles: int):
+        """Calculates the number of expected tiles we would fetch.
+
+        If this is greater than the number of max tiles, raise an error.
+        """
+        total_n_tiles = (yend - ystart + 1) * (xend - xstart + 1)
+        if total_n_tiles > max_tiles:
+            raise ValueError(f"Requested zoom results in {total_n_tiles} tiles."
+                             f"Max allowed tiles are {max_tiles}")
+
+    @validator('zoom_levels')
+    def validate_zoom_levels(cls, zoom_levels):
+        if zoom_levels[0] > zoom_levels[1]:
+            raise ValueError(
+                f"Order of zoom levels should be min, max. Received {zoom_levels}"
+            )
+        return zoom_levels
+
+
+class EPSGTransformer(BaseModel):
+    """Transformer class between different EPSG's. Useful when wanting to project
+    in different formats.
+
+    Requires as input a Point object.
+    """
+
+    class ProjectionTransformer(Transformer):
+        """Custom class to help represent a Transformer that will play
+        nicely with Pydantic.
+
+        Accepts a PyProj Transformer object.
+        """
+
+        @classmethod
+        def __get_validators__(cls):
+            yield cls.validate
+
+        @classmethod
+        def validate(cls, v):
+            if not isinstance(v, Transformer):
+                raise Exception("Needs to be a Transformer class")
+            return v
+
+    transform_function: Optional[ProjectionTransformer] = None
+
+    def _is_simple(self, epsg: EPSG) -> bool:
+        return epsg == EPSG.SIMPLEPIXEL
+
+    def _get_ranges(self, bounds: np.ndarray):
+        """helper function to get the range between bounds.
+        
+        returns a tuple (x_range, y_range)"""
+        x_range = np.max(bounds[:, 0]) - np.min(bounds[:, 0])
+        y_range = np.max(bounds[:, 1]) - np.min(bounds[:, 1])
+        return (x_range, y_range)
+
+    def _min_max_x_y(self, bounds: np.ndarray):
+        """returns the min x, max x, min y, max y of a numpy array
+        """
+        return np.min(bounds[:, 0]), np.max(bounds[:, 0]), np.min(
+            bounds[:, 1]), np.max(bounds[:, 1])
+
+    def geo_and_geo(self, src_epsg: EPSG, tgt_epsg: EPSG) -> None:
```

Comment:
This should be a classmethod constructor. Same with geo_and_pixel. Then you can make the transform_function a required field. And not have to raise raise Exception("No transformation has been set.").

Reply:
Maybe create 3 functions.

1. def create_geo_to_geo_transformer() -> EPSGTransformer (this is just geo_and_geo renamed and it will return the object instead of None. 
2. def create_geo_to_pixel_transformer() -> EPSGTransformer
3. def create_pixel_to_geo_transformer() -> EPSGTransformer

2 and 3 will call the same geo_and_pixel function under the hood but this way it is more clear what the function is doing. Also it gives us more flexibility in the future